File: packages/nxuskit-py/src/nxuskit/tools.py
"""Tool/function calling types for the nxusKit Python SDK.

These types define the canonical tool calling contract, matching the
OpenAI-compatible schema used across all nxusKit wrappers.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class FunctionDefinition:
    """A function that can be called by the model."""

    name: str
    description: str | None = None
    parameters: dict[str, Any] | None = None


@dataclass
class ToolDefinition:
    """A tool available for the model to call."""

    type: str = "function"
    function: FunctionDefinition = field(default_factory=FunctionDefinition)

    @staticmethod
    def create(
        name: str,
        description: str | None = None,
        parameters: dict[str, Any] | None = None,
    ) -> "ToolDefinition":
        """Create a tool definition for a function."""
        return ToolDefinition(
            type="function",
            function=FunctionDefinition(
                name=name,
                description=description,
                parameters=parameters,
            ),
        )

    def to_dict(self) -> dict[str, Any]:
        """Serialize to dict for JSON encoding."""
        result: dict[str, Any] = {"type": self.type, "function": {"name": self.function.name}}
        if self.function.description is not None:
            result["function"]["description"] = self.function.description
        if self.function.parameters is not None:
            result["function"]["parameters"] = self.function.parameters
        return result


@dataclass
class FunctionCall:
    """Function invocation details from the model."""

    name: str
    arguments: str  # JSON-encoded


@dataclass
class ToolCall:
    """A tool call requested by the model."""

    id: str
    type: str  # always "function"
    function: FunctionCall

    @staticmethod
    def from_dict(d: dict[str, Any]) -> "ToolCall":
        """Deserialize from dict."""
        return ToolCall(
            id=d["id"],
            type=d.get("type", "function"),
            function=FunctionCall(
                name=d["function"]["name"],
                arguments=d["function"]["arguments"],
            ),
        )


@dataclass
class ToolResultMessage:
    """Result of executing a tool, sent back to the model."""

    tool_call_id: str
    content: str
    role: str = "tool"

    def to_dict(self) -> dict[str, Any]:
        """Serialize to dict for JSON encoding."""
        return {
            "role": self.role,
            "tool_call_id": self.tool_call_id,
            "content": self.content,
        }


# ── Tool Choice helpers ───────────────────────────────────────────


def tool_choice_auto() -> str:
    """Return a tool_choice value of 'auto'."""
    return "auto"


def tool_choice_none() -> str:
    """Return a tool_choice value of 'none'."""
    return "none"


def tool_choice_required() -> str:
    """Return a tool_choice value of 'required'."""
    return "required"


def tool_choice_named(function_name: str) -> dict[str, Any]:
    """Return a tool_choice that forces a specific function."""
    return {
        "type": "function",
        "function": {"name": function_name},
    }

File: packages/nxuskit-go/tools.go
package nxuskit

// ToolCall represents a tool call requested by the model in a ChatResponse.
// The model returns these when finish_reason is "tool_calls".
type ToolCall struct {
	ID       string       `json:"id"`
	Type     string       `json:"type"` // always "function"
	Function FunctionCall `json:"function"`
}

// FunctionCall contains the function invocation details from the model.
type FunctionCall struct {
	Name      string `json:"name"`
	Arguments string `json:"arguments"` // JSON-encoded arguments
}

// ToolResultMessage sends a tool execution result back to the model.
// Use this as a message in the continuation ChatRequest.
type ToolResultMessage struct {
	Role       string `json:"role"`         // always "tool"
	ToolCallID string `json:"tool_call_id"` // references ToolCall.ID
	Content    string `json:"content"`      // result content (typically JSON)
}

// NewToolResultMessage creates a tool result message.
func NewToolResultMessage(toolCallID, content string) ToolResultMessage {
	return ToolResultMessage{
		Role:       "tool",
		ToolCallID: toolCallID,
		Content:    content,
	}
}
